README
self.pm version 0.0.1

self.pm provides 'self' and 'args' methods for your Perl OO modules, so you don't need to
say:

    my $self = shift;

This is no source filter.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

DEPENDENCIES

None.


COPYRIGHT AND LICENCE

Copyright (C) 2007, Kang-min Liu

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for self

- No longer depends on Devel::Declare
- Migrate test code to Test::V2

0.35 2016-02-01T08:20:44+0100
- remove self::implicit -- "self" already has the functionality.
- Fix code injection under perl 5.18, 5.20, and 5.22.

0.34
- fix deps
 
0.33
- license changed to MIT.
- support 'no self'

0.32
- fix deps

0.31
- Make it parses sub declaration when '{' is not in the same line as 'sub'

0.30
- Here comes the B:: -based implementation that gives you $self and @args

0.15
- More tests to proof.
- Documentation correction.

0.14
- Emergency fix for wrong depency. :(

0.13
- Make it work inside 'eval'.

0.12
- requires Perl v5.6.0 instead of v5.8.0
- Use Sub::Exporter instead of Exporter::Lite, this provide
  a 'renaming' feature, which makes many sense for this module.
- Using 'self' in eval block or statement is not working, marked as
  TODO in tests.
- All these are reported by MARKSTOS++, thanks a lot!.

0.10  Sun Oct 28 19:23:14 2007
       Initial release.



t/lib/Counter.pm
use strict;
use warnings;

package Counter;

use self;

sub new {
    my $class = shift;
    return bless {
        v => 0
    }, $class;
}

sub set {
    my ($v) = args;
    self->{v} = $v;
}

sub out {
    self->{v};
}

sub inc {
    self->{v}++;
}

1;


t/lib/EvalTest.pm
use strict;
use warnings;
package EvalTest;

use self;

sub new {
    return bless {}, self;
}

sub in {
    my ($n) = args;
    self->{n} = $n;
}

sub out {
    return self->{n}
}

sub out2 {
    return eval {
        return self->{n}
    };
}

sub out3 {
    return eval "self->{n}";
}

1;


t/lib2/Selfvar.pm
use strict;

package Selfvar;
use self;

sub new { bless {}, shift }

sub pet {
    if ($args[0]) {
        $self->{pet} = $args[0];
    }

    return $self->{pet};
}

   sub echo {
       return $self;
   }

1;


Makefile.PL
use strict;
use warnings;

use ExtUtils::MakeMaker;

WriteMakefile(
    NAME => "self",
    VERSION_FROM => "lib/self.pm",
    ABSTRACT_FROM => "lib/self.pm",
    AUTHOR => 'Kang-min Liu <gugod@gugod.org>',
    PREREQ_PM => {
        'PadWalker'        => '1.93',
        'Devel::Declare'   => '0.003004',
        'B::OPCheck'       => '0.27',
        'B::Hooks::Parser' => '0.08',
        'Sub::Exporter'    => '0',
    },
    BUILD_REQUIRES => {
        'Test2::V0'  => '0.000139',
    },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {
            repository => {
                type => 'git',
                url => 'git://github.com/gugod/self.git',
                web => 'https://github.com/gugod/self',
            }
        }
    },
    dist => {
        COMPRESS => "zopfli -i50",
    },
);


lib/self/implicit.pm
package self::implicit;
use 5.006;
use strict;
use warnings;

use Devel::Declare ();
use B::Hooks::Parser;
use PadWalker qw(peek_my);

our $VERSION = '0.33';

my $NO_SELF;

sub import {
    my ($class) = @_;
    my $caller = caller;

    B::Hooks::Parser::setup();

    my $linestr = B::Hooks::Parser::get_linestr();
    my $offset  = B::Hooks::Parser::get_linestr_offset();
    substr($linestr, $offset, 0) = 'use PadWalker ();use B::OPCheck const => check => \&self::implicit::_check;';
    B::Hooks::Parser::set_linestr($linestr);
}

# This routine is almost the same as self::_check, the only difference is the injected $code.
sub _check {
    my $op = shift;
    my $caller = caller;
    return if $NO_SELF;
    return unless ref($op->gv) eq 'B::PV';

    my $linestr = B::Hooks::Parser::get_linestr;
    my $offset  = B::Hooks::Parser::get_linestr_offset;

    my $code = q{if(ref($_[0]) ne __PACKAGE__){my $x = PadWalker::peek_my(1)->{'$self'};unshift @_,$$x if $x;};my($self,@args)=@_;};

    if (substr($linestr, $offset, 3) eq 'sub') {
        my $line = substr($linestr, $offset);
         if ($line =~ m/^sub\s.*{ /x ) {
            if (index($line, "{$code") < 0) {
                substr($linestr, $offset + index($line, '{') + 1, 0) = $code;
                B::Hooks::Parser::set_linestr($linestr);
            }
        }
    }

    # This elsif block handles:
    # sub foo
    # {
    # ...
    # }
    elsif (index($linestr, 'sub') >= 0) {
        $offset += Devel::Declare::toke_skipspace($offset);
        if ($linestr =~ /(sub.*?\n\s*{)/) {
            my $pos = index($linestr, $1);
            if ($pos + length($1) - 1 == $offset) {
                substr($linestr, $offset + 1, 0) = $code;
                B::Hooks::Parser::set_linestr($linestr);
            }
        }
    }

}

sub unimport {
    my ($class) = @_;
    my $caller = caller;
    $NO_SELF = 1;
}

1;

__END__

=head1 NAME

self::implicit - provides '$self' in OO code, in an implicit way.

=head1 SYNOPSIS

    package MathGuy;
    use self::implicit;

    # Write a normal constructor
    sub new {
        my ($class, $number) = @_;
        return bless({ number => $number }, $class);
    }

    # $self is now a special variable
    sub single() { $self->{number} }

    # It does not have to be said for method invacation
    sub double() { single * 2      }

    # '@args' too
    sub add {
        my ($n) = @args;
        return single + $n;
    }

    package main;
    my $tom = MathGuy->new(10);

    print $tom->double; #=> 20
    print $tom->add(3); #=> 13

=head1 DESCRIPTION

This module adds C<$self> and C<@args> variables in your code like
C<self.pm>, and it also eliminate the need to say '$self->' in your
code.  Normal function invocation will become method invocation, if
there is a C<$self> variable in the caller context.

=head1 AUTHOR

Kang-min Liu  C<< <gugod@gugod.org> >>

=head1 LICENCE AND COPYRIGHT

See C<self.pm>

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


inc/Module/Install/Makefile/Name.pm
#line 1
package Module::Install::Makefile::Name;

use Module::Install::Base;
@ISA = qw(Module::Install::Base);

$VERSION = '0.67';

use strict;

sub determine_NAME {
    my $self = shift;
    my @modules = glob('*.pm');

    require File::Find;
    File::Find::find( sub {
        push @modules, $File::Find::name if /\.pm/i;
    }, 'lib');

    if (@modules == 1) {
        local *MODULE;
        open MODULE, $modules[0] or die $!;
        while (<MODULE>) {
            next if /^\s*(?:#|$)/;
            $self->module_name($1) if /^\s*package\s+(\w[\w:]*)\s*;\s*$/;
            last;
        }
    }

    return if $self->module_name;

    my $name = MM->guess_name or die <<"END_MESSAGE";
Can't determine a NAME for this distribution.
Please use the 'name' function in Makefile.PL.
END_MESSAGE

    $name =~ s/-/::/g;
    $self->module_name($name);
    return $name;
}

1;


lib/self.pm
use strict;
use warnings;

package self;
use 5.006;

our $VERSION = '0.35';
use Sub::Exporter;

use B::Hooks::Parser;

my $NO_SELF;

sub import {
    my ($class) = @_;
    my $caller = caller;

    B::Hooks::Parser::setup();

    my $linestr = B::Hooks::Parser::get_linestr();
    my $offset  = B::Hooks::Parser::get_linestr_offset();
    substr($linestr, $offset, 0) = 'use B::OPCheck const => check => \&self::_check;';
    B::Hooks::Parser::set_linestr($linestr);

    my $exporter = Sub::Exporter::build_exporter({
        into_level => 1,
        exports => [qw(self args)],
        groups  => { default =>  [ -all ] }
    });
    $exporter->(@_);
}

sub unimport {
    my ($class) = @_;
    my $caller = caller;
    $NO_SELF = 1;
}

sub _check {
    my $op = shift;
    my $caller = caller;
    return if $NO_SELF;
    return unless ref($op->gv) eq 'B::PV';

    my $linestr = B::Hooks::Parser::get_linestr;
    my $offset  = B::Hooks::Parser::get_linestr_offset;
    my $line = substr($linestr, $offset);

    my $code = 'my($self,@args)=@_;';

    # This cover cases like:
    #     sub foo { ... }
    # Offset is at the first '{' because subroutine name is also a "const"
    if (substr($linestr, $offset, 1) eq '{') {
        if (substr($linestr, 0, $offset) =~ m/sub\s\S+\s*\z/x ) {
            if (index($line, "{$code") < 0) {
                substr($linestr, $offset + 1, 0) = $code;
                B::Hooks::Parser::set_linestr($linestr);
            }
        }
    }
    elsif (substr($linestr, $offset, 3) eq 'sub') {
        if ($line =~ m/^sub\s.*{ /x ) {
            if (index($line, "{$code") < 0) {
                substr($linestr, $offset + index($line, '{') + 1, 0) = $code;
                B::Hooks::Parser::set_linestr($linestr);
            }
        }
    }

    # This elsif block handles:
    # sub foo
    # {
    # ...
    # }
    elsif (index($linestr, 'sub') >= 0) {
        $offset += B::Hooks::Toke::skipspace($offset);
        if ($linestr =~ /(sub.*?\n\s*{)/) {
            my $pos = index($linestr, $1);
            if ($pos + length($1) - 1 == $offset) {
                substr($linestr, $offset + 1, 0) = $code;
                B::Hooks::Parser::set_linestr($linestr);
            }
        }
    }
}

sub _args {
    my $level = 2;
    my @c = ();
    package DB;
    @c = caller($level++)
        while !defined($c[3]) || $c[3] eq '(eval)';
    return @DB::args;
}

sub self {
    (_args)[0];
}

sub args {
    my @a = _args;
    return @a[1..$#a];
}

1;

__END__

=head1 NAME

self - provides '$self' in OO code.

=head1 VERSION

This document describes self version 0.30.

=head1 SYNOPSIS

    package MyModule;
    use self;

    # Write constructor as usual
    sub new {
        return bless({}, shift);
    }

    # '$self' is special now.
    sub foo {
        $self->{foo}
    }

    # '@args' too
    sub set {
        my ($foo, $bar) = @args;
        $self->{foo} = $foo;
        $self->{bar} = $bar;
    }

=head1 DESCRIPTION

This module adds C<$self> and C<@args> variables in your code. So you
don't need to say:

    my $self = shift;

The provided C<$self> and C<@args> are lexicals in your sub, and it's
always the same as saying:

    my ($self, @args) = @_;

... in the first line of sub.

However it is not source filtering, but compile-time code
injection. For more info about code injection, see L<B::Hooks::Parser>.

It also exports a C<self> and a C<args> functions. Basically C<self> is just
equal to C<$_[0]>, and C<args> is just C<$_[1..$#_]>.

For convienence (and backward compatibility), these two functions
are exported by default. If you don't want them to be exported, you
need to say:

    use self ();

Since self.pm uses L<Sub::Exporter>, the exported <self> funciton
can be renamed:

    use self self => { -as => 'this' };

For more information, see L<Sub::Exporter>.

It is recommended to use variables instead, because it's much much
faster. There's a benchmark program under "example" directory compare
them: Here's one example run:

    > perl -Ilib examples/benchmark.pl
              Rate  self $self
    self   46598/s    --  -92%
    $self 568182/s 1119%    --

=head1 INTERFACE

=over

=item $self, or self

Return the current object.

=item @args, or args

Return the argument list.

=back

=head1 CONFIGURATION AND ENVIRONMENT

self.pm requires no configuration files or environment variables.

=head1 DEPENDENCIES

C<B::OPCheck>, C<B::Hooks::Parser>, C<Sub::Exporter>

=head1 INCOMPATIBILITIES

None reported.

=head1 BUGS AND LIMITATIONS

In some cases, C<$self> and C<@args> may failed to be injected.

If you're using 0.30, please ensure that your sub declaration has its
'{' at the same line like this:

    sub foo {
    }

Also it's ok to have the entire sub in one line:

    sub foo { }

Please upgrade to 0.31 if you prefer this style of code:

   sub foo
   {
       $self;
   }

Extra spaces around sub declarations are handled as much as possible,
if you found any cases that it failed to work, please send me bug
reports with your test cases.

It does not work on methods generated in runtime. Remember, it's a
compile-time code injection. For those cases, use C<self> function
instead.

Please report any bugs or feature requests to C<bug-self@rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org>.

=head1 AUTHOR

Kang-min Liu  C<< <gugod@gugod.org> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2021 Kang-min Liu C<< <gugod@gugod.org> >>.

This is free software, licensed under:

    The MIT (X11) License

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.
